app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt

```diff
@@ -0,0 +1,208 @@
+            }
+
+            override fun surfaceDestroyed(holder: SurfaceHolder) {
+                camera.stop()
+            }
+        })
+
+        barcodeDetector.setProcessor(object : Detector.Processor<Barcode> {
+            override fun release() {
+                Toast.makeText(applicationContext, "Scanner has been closed", Toast.LENGTH_SHORT).show()
+            }
+
+            override fun receiveDetections(detections: Detector.Detections<Barcode>) {
+                solveDetection(detections)
+            }
+        })
+    }
+
+    override fun onDestroy() {
+        super.onDestroy()
+        // avoid error using non initialized property
+        if(::camera.isInitialized) {
+            camera.stop()
+        }
+    }
+
+    /*
+     * This section is related to permission request
+     */
+
+    /**
+     * Ask for camera permission
+     */
+    private fun askForCameraPermission() {
+        ActivityCompat.requestPermissions(
+            this@QrScanningActivity,
+            arrayOf(android.Manifest.permission.CAMERA),
+            requestCodeCameraPermission
+        )
+    }
+
+    override fun onRequestPermissionsResult(
+        requestCode: Int,
+        permissions: Array<out String>,
+        grantResults: IntArray
+    ) {
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
+        if (requestCode == requestCodeCameraPermission && grantResults.isNotEmpty()) {
+            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                // If permission are granted restart the activity to make sure they are taken into account
+                val intentNew =  Intent(this, QrScanningActivity::class.java)
+                intentNew.putExtra("uidList", uidList)
+                startActivity(intentNew)
+                finish()
+            } else {
+                // Camera permission not granted, come back to previous activity
+                Toast.makeText(applicationContext, "Camera Permission should be granted for that feature", Toast.LENGTH_LONG).show()
+                val returnInt = Intent(this@QrScanningActivity, SearchUserActivity::class.java)
+                startActivity(returnInt)
+                finish()
+            }
+        }
+    }
+
+    /*
+     * This section is related to the handler when code is detected
+     */
+
+    private fun solveDetection(detections: Detector.Detections<Barcode>) {
+        val barcodes = detections.detectedItems
+        if (isTest || barcodes.size() == 1) {
+            if (!isTest) {
+                scannedValue = barcodes.valueAt(0).rawValue
+            }
+            runOnUiThread {
+                camera.stop()
+                val endInt = Intent(this@QrScanningActivity, SearchUserActivity::class.java)
+                if (isTest || scannedValue in uidList){
+                    if (!isTest) {
+                        usersRepo.setSubFieldValue(FireBaseAuthenticator.getCurrentUID(),"friends", scannedValue, true)
```

Comment:
do you wanna try and mock the usersRepo?

Reply:
Maybe not this week, already took so much time to do it. But I can take care of that next week. Depends which task I am taking